### DISCUSSION TITLE
Create Hall of fame manually

### DISCUSSION THREAD

**Discussion by DenisSvirin** [MilesCranmer/PySR#443] [General]
Can you, please, provide an example of how to create Hall of fame manually ?

I ve created an expression before assigning it to `self.raw_julia_state_`:

```
           e = SymbolicRegression.Node(val=3.2)
            x1 = SymbolicRegression.Node(feature=1)
            xsi = SymbolicRegression.Node(val=1.224)
            one = SymbolicRegression.Node(val=1)
            two = SymbolicRegression.Node(val=2)
            mtwo = SymbolicRegression.Node(val=-2)

            r_cut = SymbolicRegression.Node(val=2.556)
            q = SymbolicRegression.Node(val=2.27)

            A = SymbolicRegression.Node(val=0.0855)
            p = SymbolicRegression.Node(val=-10.960)
            tree_rep = SymbolicRegression.Node(3,
            SymbolicRegression.Node(5, xsi, two),SymbolicRegression.Node(5,
             e, SymbolicRegression.Node(3,
              SymbolicRegression.Node(3,
               SymbolicRegression.Node(2,
                SymbolicRegression.Node(4,
                 x1, r_cut), one), q), mtwo)))

            tree_att = SymbolicRegression.Node(3,
            A, SymbolicRegression.Node(5,
             e, SymbolicRegression.Node(3,
              SymbolicRegression.Node(2,
               SymbolicRegression.Node(4,
                 x1, r_cut), one), p)))
            tree = SymbolicRegression.Node(2, tree_rep, SymbolicRegression.Node(5, tree_att, two))

```
(May be there more efficient way to do it)
But im struggling to put it in hof.
I planned to create hof and pass it here as `self.raw_julia_state_ `: [sr.py](https://github.com/MilesCranmer/PySR/blob/57dd7d2f9a2b214599e482cd7335ab0e815e8f1c/pysr/sr.py#L1800)

**Comment by MilesCranmer** [comment]
This will be easier once this PR is finished: https://github.com/MilesCranmer/PySR/pull/429 since it will store the `.sr_options` for use in creating a hall of fame object. I guess we also need to store the `.sr_dataset` object, but that will be trickier to do generally.

In Julia, the way I would do it is

```julia
using SymbolicRegression

e = Node(val=3.2)
x1 = Node(feature=1)
xsi = Node(val=1.224)
one = Node(val=1)
two = Node(val=2)
mtwo = Node(val=-2)

r_cut = Node(val=2.556)
q = Node(val=2.27)

A = Node(val=0.0855)
p = Node(val=-10.960)

options = Options(binary_operators=[+, -, *, /], unary_operators=[cos, exp])
# ^ Default options will automatically create the associated operators for use on `Node` directly
dataset = Dataset(X, y)

# Create data:
tree1 = exp(e + x1)
tree2 = cos(A + exp(p * x1))

# Format to PopMember:
member1 = PopMember(dataset, tree1, options)
member2 = PopMember(dataset, tree2, options)

# Create empty hall of fame:
hof = HallOfFame(options, Float32, Float32)

for member in (member1, member2)
    hof.members[member.complexity] = member
    hof.exists[member.complexity] = true
end
```

then the `hof` will be ready.

**Reply by DenisSvirin** [reply]
I pasted code here (https://github.com/MilesCranmer/SymbolicRegression.jl/blob/141987a3587978039712f010172b142ad09d4757/src/SymbolicRegression.jl#L660)
But my I have dataset with T = Float32 and Node with T = Float64, which gives an error.
As , Im quite incompetent in Julia,  could you, please, explain where I can promote either of these things to identical type?

**Reply by MilesCranmer** [reply]
Oh, then you probably want to make the trees all `Node{Float32}`. You can do that with

```julia
e = Node(val=3.2f0)
x1 = Node(Float32, feature=1)
xsi = Node(val=1.224f0)
one = Node(val=1f0)
two = Node(val=2f0)
mtwo = Node(val=-2f0)

r_cut = Node(val=2.556f0)
q = Node(val=2.27f0)

A = Node(val=0.0855f0)
p = Node(val=-10.960f0)
```

or alternatively just do `convert(Node{Float32}, tree)` on the trees.

**Reply by DenisSvirin** [reply]
That works, thanks.
But I think the default value for `deterministic` he should be `false` (here: https://github.com/MilesCranmer/SymbolicRegression.jl/blob/141987a3587978039712f010172b142ad09d4757/src/PopMember.jl#L95)
